flocker/node/test/test_loop.py

```diff
@@ -0,0 +1,583 @@
+        """
+        A FSM doing a convergence iteration does another iteration when
+        applying changes is done.
+        """
+        local_state = object()
+        local_state2 = object()
+        configuration = object()
+        state = object()
+        action = ControllableAction(succeed(None))
+        action2 = ControllableAction(Deferred())
+        deployer = ControllableDeployer(
+            [succeed(local_state), succeed(local_state2)],
+            [action, action2])
+        client = self.successful_amp_client([local_state, local_state2])
+        loop = build_convergence_loop_fsm(deployer)
+        loop.receive(_ClientStatusUpdate(
+            client=client, configuration=configuration, state=state))
+        # Calculating actions happened, result was run... and then we did
+        # whole thing again:
+        self.assertEqual((deployer.calculate_inputs, client.calls),
+                         ([(local_state, configuration, state),
+                           (local_state2, configuration, state)],
+                          [(NodeStateCommand, dict(node_state=local_state)),
+                           (NodeStateCommand, dict(node_state=local_state2))]))
+
+    def test_convergence_status_update(self):
+        """
+        A FSM doing convergence that receives a status update stores the
+        client, desired configuration and cluster state, which are then
+        used in next convergence iteration.
+        """
+        local_state = object()
+        local_state2 = object()
+        configuration = object()
+        state = object()
+        action = ControllableAction(Deferred())
+        action2 = ControllableAction(Deferred())
+        deployer = ControllableDeployer(
+            [succeed(local_state), succeed(local_state2)],
+            [action, action2])
+        client = self.successful_amp_client([local_state])
+        loop = build_convergence_loop_fsm(deployer)
+        loop.receive(_ClientStatusUpdate(
+            client=client, configuration=configuration, state=state))
+
+        # Calculating actions happened, action is run, but waits for
+        # Deferred to be fired... Meanwhile a new status update appears!
+        client2 = self.successful_amp_client([local_state2])
+        configuration2 = object()
+        state2 = object()
+        loop.receive(_ClientStatusUpdate(
+            client=client2, configuration=configuration2, state=state2))
+        # Action finally finishes, and we can move on to next iteration,
+        # which happens with second set of client, desired configuration
+        # and cluster state:
+        action.result.callback(None)
+        self.assertEqual(
+            (deployer.calculate_inputs, client.calls, client2.calls),
+            ([(local_state, configuration, state),
+              (local_state2, configuration2, state2)],
+             [(NodeStateCommand, dict(node_state=local_state))],
+             [(NodeStateCommand, dict(node_state=local_state2))]))
```

Comment:
:question: Something about testing that
- if multiple status updates are received before an iteration is done, that last result is used
- the second iteration doesn't start until the first is done.